### PR TITLE
Make gnocchi-api wrap uWSGI

### PR DESCRIPTION
**Pull request by jd**
As the wise Mehdi once said: "nobody reads documentation".
Everybody wants to run a daemon for the API server and everybody wants to run
`gnocchi-api'. Then let's provide it as a wrapper around uwsgi, so it's a fast
and performant HTTP by default.

### CODE REVIEW

**Comment by jd** [comment]
Pull-request updated, HEAD is now 8ae822819a727fc233d43684dc40995dc885727a

**Comment by jd** [comment]
Pull-request updated, HEAD is now f015763504133e1cb3f5cf2f09828039648fb1f3

**Comment by jd** [comment]
Pull-request updated, HEAD is now ec9993d8191e86eaafcb42f72a558aff115adcf7

**Comment by jd** [comment]
Pull-request updated, HEAD is now b4712353fd4a1187e9a144d89da8e47d7711d667

**Comment by jd** [comment]
Pull-request updated, HEAD is now f7f449535fad8b5999977bd73e89e67f302a37a6

File: doc/source/operating.rst [hunk 1]
```diff
@@ -37,6 +38,9 @@ The following uwsgi configuration file can be used::
   buffer-size = 65535
   lazy-apps = true
```
Comment (sileht) [inline doc/source/operating.rst]: We should add the keepalive workaround until we understand why it doesn't work.

File: gnocchi/cli.py [hunk 2]
```diff
@@ -309,6 +314,55 @@ def metricd_tester(conf):
...
+    return os.execl(
+        uwsgi, uwsgi,
+        "--http", "%s:%d" % (conf.host or conf.api.host,
+                             conf.port or conf.api.port),
+        "--master",
+        "--enable-threads",
+        "--die-on-term",
+        "--processes", str(math.floor(workers * 1.5)),
+        "--threads", str(workers),
+        "--lazy-apps",
```
Comment (sileht) [inline gnocchi/cli.py]: Same here.

**Comment by jd** [comment]
Pull-request updated, HEAD is now 6fcbbd8425620c4d0d45e94740ca0095d2d8bda8

**Comment by jd** [comment]
Pull-request updated, HEAD is now 87aa4aa4a9fdd0d6bb49e837351fbc85de3f3983

File: doc/source/operating.rst [hunk 3]
```diff
@@ -13,14 +13,15 @@ To run Gnocchi, simply run the HTTP server and metric daemon:
...
-have, usually around 1.5 × number of CPU. If one server is not enough, you can
-spawn any number of new API server to scale Gnocchi out, even on different
-machines.
+To run Gnocchi API, you can use the provided `gnocchi-api`. It wraps around
+`uwsgi` – makes sure that `uwsgi`_ is installed. If one Gnocchi API server is
+not enough, you can spawn any number of new API server to scale Gnocchi out,
+even on different machines.
 
-The following uwsgi configuration file can be used::
+Since Gnocchi API tier runs using WSGI, it can alternatively be ran using
```
Comment (chungg) [inline doc/source/operating.rst]: s/ran/run/

File: doc/source/operating.rst [hunk 4]
```diff
@@ -13,14 +13,15 @@ To run Gnocchi, simply run the HTTP server and metric daemon:
...
+`Apache httpd`_ and `mod_wsgi`_, or any other HTTP daemon. If you want to
+deploy using uwsgi `yourself`, the following uwsgi configuration file can be
```
Comment (chungg) [inline doc/source/operating.rst]: why is yourself is quotations?

File: doc/source/operating.rst [hunk 5]
```diff
@@ -36,6 +37,10 @@ The following uwsgi configuration file can be used::
   plugins = python
   buffer-size = 65535
   lazy-apps = true
+  add-header = Connection: close
+
+You should configure the number of process and threads according to the number
+of CPU you have, usually around 1.5 × number of CPU.
```
Comment (chungg) [inline doc/source/operating.rst]: isn't really clear. should the processes be 1.5x or the threads? or both? for me, the threads didn't really do anything

File: doc/source/operating.rst (hunk 3 above)
Comment (jd) [inline doc/source/operating.rst]: done

File: doc/source/operating.rst (hunk 4 above)
Comment (jd) [inline doc/source/operating.rst]: LOL I don't know! fixed.

File: doc/source/operating.rst (hunk 5 above)
Comment (jd) [inline doc/source/operating.rst]: removed threads then!

**Comment by jd** [comment]
Pull-request updated, HEAD is now 5a10a92d24ca30f482967e7dbb57339876ae5167

**Pull request merged**